db/clientlist_virtual_reader.cc

```diff
@@ -0,0 +1,237 @@
+            });
+        });
+    });
+}
+
+// Clusters clients by IP and sorts them by port number within clusters.
+future<std::vector<system_keyspace::client_data>>
+clientlist_mutation_reader::filter_and_sort_by_port(std::vector<system_keyspace::client_data>&& clients) const {
+    std::vector<system_keyspace::client_data> sorted_clients;
+    sorted_clients.reserve(clients.size());
+    
+    std::unordered_map<
+            decltype(system_keyspace::client_data::ip),
+            std::vector<system_keyspace::client_data>> pk_clusters;
+    
+    for (auto&& client : clients) {
+        const auto pk = partition_key::from_single_value(*_schema, inet_addr_type->decompose(client.ip));
+        const dht::token calculated_token = dht::global_partitioner().get_token(*_schema, pk);
+        if (dht::shard_of(calculated_token) != engine().cpu_id()) {
+            continue;
+        }
+        for (auto& range : _slice.row_ranges(*_schema, pk)) {
+            const auto ck_bytes = data_type_for<decltype(system_keyspace::client_data::port
+                    )>()->decompose(client.port);
+            const auto ck_prefix = clustering_key_prefix::from_single_value(*_schema, ck_bytes);
+
+            if (range.contains(ck_prefix, ck_comparator(_schema))) {
+                pk_clusters[client.ip].emplace_back(std::move(client));
+                break;
+            }
+        }
+    }
+
+    for (auto& [ip, cluster] : pk_clusters) {
+        if (_slice.options.contains(query::partition_slice::option::reversed)) {
+            std::sort(
+                    cluster.begin(), cluster.end(),
+                    [] (const auto& client1, const auto& client2) { return client1.port > client2.port; });
+        } else {
+            std::sort(
+                    cluster.begin(), cluster.end(),
+                    [] (const auto& client1, const auto& client2) { return client1.port < client2.port; });
+        }
+        sorted_clients.insert(
+                sorted_clients.end(),
+                make_move_iterator(cluster.begin()),
+                make_move_iterator(cluster.end()));
+    }
+    return make_ready_future<std::vector<system_keyspace::client_data>>(std::move(sorted_clients));
+}
+
+future<> clientlist_mutation_reader::maybe_setup_mutation_fragmentator() {
+    if (_mutation_fragmentator_opt) {
+        return make_ready_future<>();
+    }
+
+    return smp::submit_to(0, [] {
+        // It must be run on shard 0. Shards with id>=1 see: get_dist_cql_server()==nullptr
+        return std::ref(*service::get_local_storage_service().get_dist_cql_server());
+    }).then([] (distributed<cql_transport::cql_server>& distributed_cql_server) {
+        return get_all_clients_data(distributed_cql_server);
+    }).then([this] (auto&& client_data_vec) {
+        return filter_and_sort_by_port(std::move(client_data_vec)).then([this] (auto&& sorted_client_data_vec) {
```

Comment:
couldn't this `then` be attached one level higher so we don't have this nesting?
also, this lambda calling `filter_and_sort_by_port` is probably not needed, similarly to `get_all_clients_data` above (I'm 100% not sure about that though).
If I'm right, you could do
`then(get_all_clients_data).then(filter_and_sort_by_port).then([this], (... sorted_client_data_vec) { ... })`

Reply:
Also don't take `sorted_client_data_vec` by forwarding reference (after fixing Piotr's comment, by rvalue reference).
Just take it by value.